[addons] add gui control classes to new addon interface

AlwinEsch opened the pull request:
## Description
<!--- Provide a general summary of your change in the Title above -->
This change alow addons to edit all used controls on his processed window.

For the moment is this mostly the last request to create the global addon
interface function.

Most further requests are to change addon types iself to new style.

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ksooo commented on xbmc/addons/interfaces/GUI/controls/SettingsSlider.cpp:
```diff
@@ -0,0 +1,271 @@
+    return;
+  }
+
+  // create message
```
Remove this comment, please. It adds no value.

ksooo commented on xbmc/addons/interfaces/GUI/controls/SettingsSlider.cpp:
```diff
@@ -0,0 +1,271 @@
+  CGUIMessage msg(GUI_MSG_LABEL_SET, control->GetParentID(), control->GetID());
+  msg.SetLabel(text);
+
+  // send message
```
Remove this comment, please. It adds no value.

ksooo commented on xbmc/addons/interfaces/GUI/controls/Slider.cpp:
```diff
@@ -0,0 +1,266 @@
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSlider::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return 0;
```
"return -1;" to indicate error, like done in all (?) other methods returning int?

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISpinControlEx* control = static_cast<CGUISpinControlEx*>(handle);
+  if (!addon || !control)
```
check "text" for not being nullptr?

AlwinEsch replied:
Like above, equal to my head and was lost in memory.

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+    return;
+  }
+
+  // create message
```
remove this

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+  CGUIMessage msg(GUI_MSG_LABEL_SET, control->GetParentID(), control->GetID());
+  msg.SetLabel(text);
+
+  // send message
```
remove this

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+  control->SetType(type);
+}
+
+void Interface_GUIControlSpin::add_string_label(void* kodiBase, void* handle, const char* label, const char* value)
```
why "add", not "set"? if only one label is possible, I would use "set".

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+  return strdup(control->GetStringValue().c_str());
+}
+
+void Interface_GUIControlSpin::add_int_label(void* kodiBase, void* handle, const char* label, int value)
```
"add" vs. "set"?

ksooo commented on xbmc/addons/interfaces/GUI/controls/Spin.cpp:
```diff
@@ -0,0 +1,296 @@
+  {
+    CLog::Log(LOGERROR, "Interface_GUIControlSpin::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, control, addon ? addon->ID().c_str() : "unknown");
+    return 0;
```
"return -1;"?

AlwinEsch replied:
Not sure what's the best, has set it as 0 to have equal with float error return and is normally the default value if nothing becomes set before.

ksooo commented on xbmc/addons/interfaces/GUI/controls/TextBox.cpp:
```diff
@@ -0,0 +1,142 @@
+    return;
+  }
+
+  // create message
```
this should go

ksooo commented on xbmc/addons/interfaces/GUI/controls/TextBox.cpp:
```diff
@@ -0,0 +1,142 @@
+  CGUIMessage msg(GUI_MSG_LABEL_SET, control->GetParentID(), control->GetID());
+  msg.SetLabel(text);
+
+  // send message
```
remove me!

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/controls/Edit.h:
```diff
@@ -0,0 +1,278 @@
+    ///
+    virtual ~CEdit()
+    {
+    }
```
One of my favorites (as you know): "virtual ~CEdit() = default;" ? Here and everywhere in this PR. ;-)

AlwinEsch replied:
New problem becomes changed, is still new for me :-)

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/controls/Rendering.h:
```diff
@@ -0,0 +1,215 @@
+        m_interface->kodi_gui->control_rendering->set_callbacks(m_interface->kodiBase, m_controlHandle, this,
+                                              OnCreateCB, OnRenderCB, OnStopCB, OnDirtyCB);
+      else
+        kodi::Log(ADDON_LOG_FATAL, "kodi::gui::controls::CRendering can't create control class from Kodi !!!");
```
Use __FUNCTION__ here and at the other places where you hard coded the funtion name for logging?

ksooo reviewed (changes requested):
5000+ lines of code to review. That was hard candy. You ow me a beer at next Devcon, @AlwinEsch :-)

AlwinEsch commented on xbmc/addons/interfaces/GUI/controls/SettingsSlider.cpp:
```diff
@@ -0,0 +1,271 @@
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUISettingsSliderControl* control = static_cast<CGUISettingsSliderControl*>(handle);
+  if (!addon || !control)
```
Not really, this need it and I need a clearer head ;-).

AlwinEsch commented on xbmc/addons/interfaces/GUI/controls/Label.cpp:
```diff
@@ -0,0 +1,96 @@
+  CGUIMessage msg(GUI_MSG_LABEL_SET, control->GetParentID(), control->GetID());
+  msg.SetLabel(label);
+
+  // send message
```
Becomes removed. Always this c&p things :-D 
https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/legacy/Control.cpp#L139-L147

AlwinEsch commented on xbmc/addons/interfaces/GUI/controls/Image.cpp:
```diff
@@ -0,0 +1,90 @@
+  control->SetVisible(visible);
+}
+
+void Interface_GUIControlImage::set_filename(void* kodiBase, void* handle, const char* filename, const bool use_Cache)
```
Thought to have all removed, thank you. It flies out.

With the "use_cache" has I used to have more equal to "C" name style (has wrongly set one character as uppercase).

Only leaved everywhere the "kodiBase" as sign for me that a C++ base is behind.

AlwinEsch commented on xbmc/addons/interfaces/GUI/controls/FadeLabel.h:
```diff
@@ -0,0 +1,66 @@
+    static void set_enabled(void* kodiBase, void* handle, bool enabled);
+    static void set_selected(void* kodiBase, void* handle, bool selected);
+
+    static void add_label(void* kodiBase, void* handle, const char* label);
```
It can have more strings in a list, for them is the set_scrolling and reset also used.

ksooo replied:
Ah, okay then.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -840,6 +853,164 @@ void* Interface_GUIWindow::get_control_button(void* kodiBase, void* handle, int
+    return nullptr;
+  }
+
+  return pAddonWindow->GetAddonControl(control_id, CGUIControl::GUICONTROL_TEXTBOX, "textbox");
```
If this is not problematic with a static_cast, looks it really better.

AlwinEsch replied:
Forget my comment and becomes changed, thanks a lot!

AlwinEsch commented:
Thanks a lot and the beer is ready.

ksooo reviewed (changes requested):
Just another small nitpick, the rest is looking very good now.

AlwinEsch committed:
[addons] add gui control classes to new addon interface

This change alow addons to edit all used controls on his processed window.

For the moment is this mostly the last request to create the global addon
interface function.

Most further requests are to change addon types iself to new style.

AlwinEsch commented:
jenkins build this please

ksooo reviewed (approved):
Good to go.
